### PR TITLE
Make layer surface support popup

### PR DESCRIPTION
**Pull request by wineee**
1. configure layer surface in onEnterOutput maybe too late, Move to Component.onCompleted
2. Support popup'parent is layer surface

### CODE REVIEW

File: src/server/kernel/wxdgsurface.cpp [hunk 1]
```diff
@@ -287,6 +295,32 @@ WXdgSurface *WXdgSurface::parentXdgSurface() const
         if (!parent)
             return nullptr;
         return fromHandle(QWXdgToplevel::from(parent));
+    } else if (auto popup = d->handle->toPopup()) {
```
Comment (zccrs) [inline src/server/kernel/wxdgsurface.cpp:298]: Why

File: src/server/qtquick/private/wquickxdgshell.cpp [hunk 2]
```diff
@@ -134,7 +134,15 @@ void WXdgSurfaceItem::onSurfaceCommit()
 {
     WSurfaceItem::onSurfaceCommit();
     if (auto popup = m_surface->handle()->toPopup()) {
-        setImplicitPosition(popup->getPosition() - contentItem()->position());
+        if (m_surface->parentLayerSurface()) {
```
Comment (zccrs) [inline src/server/qtquick/private/wquickxdgshell.cpp:137]: You can do it in WXdgSurface::getPopupPosition().
And don't check its parent is WLayerSurface.

**Review by zccrs — CHANGES_REQUESTED** [review]

File: flake.nix [hunk 3]
```diff
@@ -38,6 +38,7 @@
               wayland-utils
               extra-cmake-modules
               qt6.qtwayland
+              libinput
```
[the remark this answers is not in the document]
Comment (wineee) [reply flake.nix:41]: for qwlroots submodule build

File: src/server/kernel/wxdgsurface.cpp (hunk 1 above)
Comment (wineee) [reply src/server/kernel/wxdgsurface.cpp:298]: `parentXdgSurface` of Popup is still useful in  https://github.com/vioken/waylib/blob/master/examples/tinywl/StackWorkspace.qml#L106-L117, should not remove

File: src/server/kernel/wxdgsurface.h [hunk 4]
```diff
@@ -63,6 +66,7 @@ public Q_SLOTS:
 
     bool checkNewSize(const QSize &size) override;
     void resize(const QSize &size) override;
+    void destroy();
```
Comment (zccrs) [inline src/server/kernel/wxdgsurface.h:69]: remove?

**Review by zccrs — CHANGES_REQUESTED** [review]

File: examples/tinywl/StackWorkspace.qml [hunk 5]
```diff
@@ -101,19 +102,20 @@ Item {
             property string type
 
             property alias xdgSurface: surface
-            property var xdgParent: root.getSurfaceItemFromWaylandSurface(waylandSurface.parentXdgSurface)
+            property var parentItem: root.getSurfaceItemFromWaylandSurface(waylandSurface.parentSurface)
 
-            parent: xdgParent ? xdgParent.shell : root
-            visible: xdgParent.item.effectiveVisible && waylandSurface.surface.mapped && waylandSurface.WaylandSocket.rootSocket.enabled
+            parent: parentItem ? parentItem.shell : root
+            visible:parentItem && parentItem.item.effectiveVisible
```
Comment (zccrs) [inline examples/tinywl/StackWorkspace.qml:108]: `visible: `

File: examples/tinywl/TiledWorkspace.qml [hunk 6]
```diff
@@ -9,9 +9,10 @@ import Waylib.Server
 Item {
     id: root
     required property Item activeFocusItem
-    function getXdgSurfaceFromWaylandSurface(surface) {
+    function getSurfaceItemFromWaylandSurface(surface) {
         let finder = function(props) {
-            if (props.waylandSurface === surface)
+            // surface is WToplevelSurface or WSurfce
+            if (props.waylandSurface === surface || props.waylandSurface.surface === surface)
```
Comment (zccrs) [inline examples/tinywl/TiledWorkspace.qml:15]:
````
if (!props.waylandSurface)
    return
````

File: examples/tinywl/TiledWorkspace.qml [hunk 7]
```diff
@@ -104,12 +105,21 @@ Item {
...
-                property var xdgParent: root.getXdgSurfaceFromWaylandSurface(waylandSurface.parentXdgSurface)
-
-                parent: xdgParent ? xdgParent.parent : root
-                visible: xdgParent.xdgSurface.effectiveVisible && waylandSurface.surface.mapped && waylandSurface.WaylandSocket.rootSocket.enabled
-                x: surface.implicitPosition.x + (xdgParent ? xdgParent.xdgSurface.contentItem.x : 0)
-                y: surface.implicitPosition.y + (xdgParent ? xdgParent.xdgSurface.contentItem.x : 0)
+                property var parentItem: root.getSurfaceItemFromWaylandSurface(waylandSurface.parentSurface)
+
+                parent: parentItem ? parentItem.shell : root
+                visible:parentItem && parentItem.item.effectiveVisible
```
Comment (zccrs) [inline examples/tinywl/TiledWorkspace.qml:111]: `visible: `

File: src/server/kernel/wxdgsurface.h [hunk 8]
```diff
@@ -22,6 +23,7 @@ class WAYLIB_SERVER_EXPORT WXdgSurface : public WToplevelSurface, public WObject
     Q_PROPERTY(bool isPopup READ isPopup CONSTANT)
     Q_PROPERTY(bool isResizeing READ isResizeing NOTIFY resizeingChanged FINAL)
     Q_PROPERTY(WXdgSurface* parentXdgSurface READ parentXdgSurface NOTIFY parentXdgSurfaceChanged FINAL)
+
```
Comment (zccrs) [inline src/server/kernel/wxdgsurface.h:26]: unnecessary

File: src/server/kernel/wxdgsurface.cpp [hunk 9]
```diff
@@ -324,6 +325,12 @@ WSurface *WXdgSurface::parentSurface() const
     return nullptr;
 }
 
+QPoint WXdgSurface::getPopupPosition() const
+{
+    auto *popup = handle()->toPopup();
+    return {popup->handle()->current.geometry.x, popup->handle()->current.geometry.y};
```
Comment (zccrs) [inline src/server/kernel/wxdgsurface.cpp:331]:
````
if (popup->parent is a XdgSurface)
    return popup->getPosition();
...
````

**Review by zccrs — CHANGES_REQUESTED** [review]

File: src/server/kernel/wxdgsurface.cpp (hunk 9 above)
Comment (wineee) [reply src/server/kernel/wxdgsurface.cpp:331]: need https://github.com/vioken/qwlroots/pull/218

Comment (zccrs) [reply src/server/kernel/wxdgsurface.cpp:334]: Merged

**Review by zccrs — APPROVED** [review]

File: src/server/kernel/wxdgsurface.cpp (hunk 9 above)
Comment (wineee) [reply src/server/kernel/wxdgsurface.cpp:334]: Done

**Pull request merged**